client/driver/docker_progress.go

```diff
@@ -0,0 +1,264 @@
+package driver
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/docker/docker/pkg/jsonmessage"
+	units "github.com/docker/go-units"
+)
+
+const (
+	// dockerPullActivityDeadline is the default value set in the imageProgressManager
+	// when newImageProgressManager is called
+	dockerPullActivityDeadline = 2 * time.Minute
+
+	// dockerImageProgressReportInterval is the default value set in the
+	// imageProgressManager when newImageProgressManager is called
+	dockerImageProgressReportInterval = 10 * time.Second
+)
+
+// layerProgress tracks the state and downloaded bytes of a single layer within
+// a docker image
+type layerProgress struct {
+	id           string
+	status       layerProgressStatus
+	currentBytes int64
+	totalBytes   int64
+}
+
+type layerProgressStatus int
+
+const (
+	layerProgressStatusUnknown layerProgressStatus = iota
+	layerProgressStatusStarting
+	layerProgressStatusWaiting
+	layerProgressStatusDownloading
+	layerProgressStatusVerifying
+	layerProgressStatusDownloaded
+	layerProgressStatusExtracting
+	layerProgressStatusComplete
+	layerProgressStatusExists
+)
+
+func lpsFromString(status string) layerProgressStatus {
+	switch status {
+	case "Pulling fs layer":
+		return layerProgressStatusStarting
+	case "Waiting":
+		return layerProgressStatusWaiting
```

Comment:
This one is also unused. Did we want to check if the status is `layerProgressWaiting` in the `get()` method below and return a nicer string message for it?